packages/muze-utils/src/DataSystem/selection.js

```diff
@@ -105,109 +82,162 @@ class Selection {
-            this._idMap[id] = params[0];
-            this._data.push(params[0]);
-        });
-        // reset enter selection
-        this._enterdata = [];
-        return this;
-    }
+    merge (target) {
+        const mergedObjects = new Map();
+        const mergedData = new Map();
+        let val;
+        let entries;
 
-    /**
-     * Returns a selection with exit data.
-     *
-     * @return {Selection} Instance of selection.
-     * @memberof Selection
-     */
-    exit () {
-        this._mode = 'exit';
-        // const exitdata = this._exitdata;
-        // const exitSelection = new Selection(exitdata);
-        // this._exitdata = [];
-        return this;
+        // merge Object and data present in this selection
+        entries = this._updatedata.entries();
+        while (val = entries.next().value) {
+            if (!this._exitdata.has(val[0])) {
+                mergedData.set(val[0], val[1]);
+                mergedObjects.set(val[0], this._dataObjects.get(val[0]));
+            }
+        }
+
+        // merge object from target selection
+        entries = target._updatedata.entries();
+        while (val = entries.next().value) {
+            if (!(mergedData.has(val[0]) || target._exitdata.has(val[0]))) {
+                mergedData.set(val[0], val[1]);
+                mergedObjects.set(val[0], target._dataObjects.get(val[0]));
+            }
+        }
+
+        const newSelection = new Selection(this._idGetter);
+        newSelection._updatedata = mergedData;
+        newSelection._dataObjects = mergedObjects;
+
+        return newSelection;
     }
 
     each (fn) {
-        Object.keys(this._idMap).forEach((e, i) => {
-            fn(this._idMap[e], this._dataMap[e], i);
-        });
-        return this;
+        let val;
+        const entries = this._dataObjects.entries();
+
+        while (val = entries.next().value) {
+            fn(val[1], this._updatedata.get(val[0]), val[0]);
+        }
     }
 
     map (fn) {
-        Object.keys(this._idMap).forEach((...params) => {
-            const key = params[0];
-            this._idMap[key] = fn(this._idMap[key], ...params, this._dataMap[key]);
-        });
+        const newdata = new Map();
+        let val;
+        const entries = this._dataObjects.entries();
+
+        while (val = entries.next().value) {
+            newdata.set(val[0], fn(val[1], this._updatedata.get(val[0]), val[0]));
+        }
+        const newSelection = new Selection(this._idGetter);
+        newSelection._updatedata = this._updatedata;
+        newSelection._dataObjects = newdata;
+
+        return newSelection;
+    }
+
+    remove () {
+        let currentData;
+
+        switch (this._mode) {
+        case 'enter':
+            currentData = this._enterdata;
+            break;
+        case 'exit':
+            currentData = this._exitdata;
+            break;
+        default:
+            currentData = this._updatedata;
+        }
+
+        currentData.clear();
```

Comment:
Also need to clear the `dataObjects`

Reply:
dataobjects are never created before `append or merge or map`, which in turn returns a new selection of dataobjects created from the mode of the previous.

Reply:
`selection.exit().remove()` In this case, the data array is removed but the data objects are not removed